File: 1052-grumpy-bookstore-owner/1052-grumpy-bookstore-owner.java
class Solution {
    

    public int maxSatisfied(final int[] customers, final int[] grumpy, final int minutes) {
        final int n = customers.length;
        final List<Integer> pre = new ArrayList<>();
        pre.add(0);
        int res = 0;
        for (int i = 0; i < n; i++) {
            if (grumpy[i] == 0) {
                res += customers[i];
                pre.add(pre.get(i));
            } else {
                pre.add(pre.get(i) + customers[i]);
            }
        }

        int mx = 0;
        for (int i = minutes; i <= n; i++) {
            mx = Math.max(mx, pre.get(i) - pre.get(i - minutes));
        }

        return res + mx;
    }
}

File: 1052-grumpy-bookstore-owner/1052-grumpy-bookstore-owner.py
class Solution:
    def maxSatisfied(self, customers: List[int], grumpy: List[int], minutes: int) -> int:
        n = len(customers)
        res = 0
        pre = [0]

        for i in range(n):
            if not grumpy[i]:
                res += customers[i]
                pre.append(pre[-1])
            else:
                pre.append(pre[-1] + customers[i])

        mv = 0
        for i in range(minutes, n + 1):
            mv = max(mv, pre[i] - pre[i-minutes])

        return res + mv
